frontends/common/constantFolding.cpp

```diff
@@ -35,38 +32,34 @@ const IR::Expression* DoConstantFolding::getConstant(const IR::Expression* expr)
         for (auto e : list->components)
             if (getConstant(e) == nullptr)
                 return nullptr;
-        return list;
+        return expr;
     }
     if (typesKnown) {
         auto ei = EnumInstance::resolve(expr, typeMap);
         if (ei != nullptr)
             return expr;
     }
-
     return nullptr;
 }
 
-// This has to be called from a visitor method - it calls getOriginal()
-void DoConstantFolding::setConstant(const IR::Node* node, const IR::Expression* result) {
-    LOG2("Folding " << node << " to " << result << " (" << result->id << ")");
-    auto orig = getOriginal();
-    constants.emplace(node, result);
-    constants.emplace(orig, result);
-}
-
 const IR::Node* DoConstantFolding::postorder(IR::PathExpression* e) {
     if (refMap == nullptr)
         return e;
     auto decl = refMap->getDeclaration(e->path);
     if (decl == nullptr)
         return e;
-    auto v = get(constants, decl->getNode());
-    if (v == nullptr)
-        return e;
-    setConstant(e, v);
-    if (v->is<IR::ListExpression>())
-        return e;
-    return v;
+    if (decl->is<IR::Declaration_Constant>()) {
+        auto dc = decl->to<IR::Declaration_Constant>();
```

Comment:
Stylisticly I prefer combining these two lines as

    if (auto dc = decl->to<IR::Declaration_Constant>()) {